bootstrap/api/v1alpha1/rke2config_types.go

```diff
@@ -18,47 +18,624 @@ package v1alpha1
+type Registry struct {
+
+	// Mirrors are namespace to mirror mapping for all namespaces.
+	Mirrors map[string]Mirror `yaml:"mirrors" json:"mirrors,omitempty"`
+
+	// Configs are configs for each registry.
+	// The key is the FDQN or IP of the registry.
+	Configs map[string]RegistryConfig `yaml:"configs" json:"configs,omitempty"`
+}
+
+// Mirror contains the config related to the registry mirror
+type Mirror struct {
+
+	// Endpoints are endpoints for a namespace. CRI plugin will try the endpoints
+	// one by one until a working one is found. The endpoint must be a valid url
+	// with host specified.
+	// The scheme, host and path from the endpoint URL will be used.
+	Endpoints []string `yaml:"endpoint" json:"endpoint,omitempty"`
+
+	// Rewrites are repository rewrite rules for a namespace. When fetching image resources
+	// from an endpoint and a key matches the repository via regular expression matching
+	// it will be replaced with the corresponding value from the map in the resource request.
+	Rewrites map[string]string `yaml:"rewrite" json:"rewrite,omitempty"`
+}
+
+// RegistryConfig contains configuration used to communicate with the registry.
+type RegistryConfig struct {
+	// Auth si a reference to a Secret containing information to authenticate to the registry.
+	// The Secret must provite a username attribute and a password attribute
+	//+optional
+	AuthSecret SecretRef `json:"authSecret,omitempty"`
+	// TLS is a pair of CA/Cert/Key which then are used when creating the transport
+	// that communicates with the registry.
+	//+optional
+	TLS TLSConfig `yaml:"tls" json:"tls,omitempty"`
+}
+
+// AuthConfig contains the config related to authentication to a specific registry
+// This is not used in the CRDs but is used for marshalling the registries.yaml file
+type AuthConfig struct {
+	// Username is the username to login the registry.
+	Username string `yaml:"username" json:"username,omitempty"`
+	// Password is the password to login the registry.
+	Password string `yaml:"password" json:"password,omitempty"`
+
+	// IdentityToken is used to authenticate the user and get
+	// an access token for the registry.
+	IdentityToken string `yaml:"identity_token" json:"identityToken,omitempty"`
+}
+
+// TODO: Beware here: TLSConfig is part of the Registry struct which has a double usage:
+// - Used in the RKE2Config CRD
+// - Used to generate a registries.yaml file for the RKE2 command line
+// In the CRD, CASecret attribute is a reference to a Secret in the Kube API, while in the registries.yaml CASecret is marshalled into ca_file which a file path on the system
+// This means the Registry object will need to be copied and processed before marshalling
+
+// TLSConfig contains the CA/Cert/Key used for a registry
+type TLSConfig struct {
+
+	// CASecret is a reference to a secret containing the Certificate Authority used to verify the registry's certificate
+	//+optional
+	CASecret SecretRef `json:"caSecret,omitempty"`
```

Comment:
can CASecret, CertSecret, KeySecret be stored in one secret?

Reply:
This could be, however, Kubernetes supports `TLS`-type secrets, and I thought I would suggest such a format for the user. Aggregating all certificates in one secret would need the secret to be a standard `Opaque `secret, but it is a trade-off